WordPress/Classes/Utility/DiskCache.swift

```diff
@@ -0,0 +1,273 @@
+import Foundation
+import CryptoKit
+
+/// An LRU disk cache that stores data in files on disk.
+///
+/// ``DiskCache`` uses LRU cleanup policy (least recently used items are removed
+/// first). The elements stored in the cache are automatically discarded if
+/// either *cost* or *count* limit is reached. The sweeps are performed periodically.
+///
+/// - important: It's possible to have more than one instance of ``DiskCache`` with
+/// the same path but it is not recommended.
+public actor DiskCache {
+    /// The path for the directory managed by the cache.
+    public nonisolated let rootURL: URL
+
+    /// The cache configuration.
+    public nonisolated let configuration: Configuration
+
+    public struct Configuration {
+        /// Size limit in bytes. `100 Mb` by default.
+        ///
+        /// Changes to the size limit will take effect when the next LRU sweep is run.
+        public var sizeLimit: Int = 1024 * 1024 * 100
+
+        /// When performing a sweep, the cache will remote entries until the size of
+        /// the remaining items is lower than or equal to `sizeLimit * trimRatio` and
+        /// the total count is lower than or equal to `countLimit * trimRatio`. `0.7`
+        /// by default.
+        var trimRatio = 0.7
+
+        /// The number of seconds between each LRU sweep. 30 by default.
+        /// The first sweep is performed right after the cache is initialized.
+        ///
+        /// Sweeps are performed in a background and can be performed in parallel
+        /// with reading.
+        public var sweepInterval: TimeInterval = 30
+
+        /// The delay after which the initial sweep is performed. 10 by default.
+        /// The initial sweep is performed after a delay to avoid competing with
+        /// other subsystems for the resources.
+        var initialSweepDelay: TimeInterval = 10
+
+        /// Initializes the configuration.
+        ///
+        /// - parameter sizeLimit: Size limit in bytes. `100 Mb` by default.
+        public init(sizeLimit: Int = 1024 * 1024 * 100) {
+            self.sizeLimit = sizeLimit
+        }
+    }
+
+    /// Creates a cache instance with a given `name`. The cache creates a directory
+    /// with the given `name` in a `.cachesDirectory` in `.userDomainMask`.
+    public init(name: String, configuration: Configuration = .init()) {
+        let cachesURL = DiskCache.getCachesURL()
+        self.init(url: cachesURL.appendingPathComponent(name, isDirectory: true), configuration: configuration)
+    }
+
+    private static func getCachesURL() -> URL {
+        guard let cachesURL = FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first else {
+            DDLogError("Failed to instantiate cache. Caches directory not available")
+            return URL(fileURLWithPath: "/dev/null") // This should never happen
+        }
+        return cachesURL
+    }
```

Comment:
I'm not sure if the app would delete all the files in the cache directory upon logout. If it doesn't, we may has an issue where cache saved in one account is used by another account. This of course can be mitigated by users passing an account specific value to the `name` argument. But I think it'd be good that this type has built-in support of separating account data, instead of relying on the users. Maybe we can add an account related argument to the initializer, like "accountIdentifier`?

Reply:
That's a good point. How does the Core Data based store currently solves it?

Reply:
I have a prototype with the following API:

```swift
final class DiskCache {
    func container(named name: String) -> DiskCacheContainer {
        DiskCacheContainer(rootURL: rootURL.appendingPathComponent(name, isDirectory: true), queue: queue)
    }
}

struct DiskCacheContainer {
    func setValue(...)
    func setData(...)

    // Creates a nested container
    func container(named name: String) -> DiskCacheContainer
}
```

It seems a bit overcomplicated, but I think it can work.
WDYT?

P.S. I wish the app had a solution for dependency injection that I could automatically inject a pre-configured container on login. But since it's not there, my goal was to make it easier to create a container on the fly and make it inexpensive to do so.

Reply:
I think it should be OK as is, but I will add the code to clear the cache on logout in the next PR.
I'm not sure Core Data does anything to handle it, does it?